js/source/geojson_source.js

```diff
@@ -95,82 +90,52 @@ GeoJSONSource.prototype = util.inherit(Evented, /** @lends GeoJSONSource.prototy
         this._data = data;
         this._dirty = true;
 
-        this.fire('change');
-
-        if (this.map)
-            this.update(this.map.transform);
+        this._onChange();
 
         return this;
     },
 
-    onAdd: function(map) {
-        this.map = map;
-    },
-
-    loaded: function() {
-        return this._loaded && this._pyramid.loaded();
-    },
-
-    update: function(transform) {
-        if (this._dirty) {
-            this._updateData();
-        }
-
-        if (this._loaded) {
-            this._pyramid.update(this.used, transform);
-        }
-    },
-
-    reload: function() {
-        if (this._loaded) {
-            this._pyramid.reload();
-        }
-    },
-
-    serialize: function() {
-        return {
-            type: 'geojson',
-            data: this._data
-        };
-    },
-
-    getVisibleCoordinates: Source._getVisibleCoordinates,
-    getTile: Source._getTile,
-
-    queryRenderedFeatures: Source._queryRenderedVectorFeatures,
-    querySourceFeatures: Source._querySourceFeatures,
-
-    _updateData: function() {
+    _updateData: function(cb) {
         this._dirty = false;
-        var options = {
-            tileSize: this.tileSize,
+        var options = util.extend({
             source: this.id,
-            geojsonVtOptions: this.geojsonVtOptions,
-            cluster: this.cluster,
-            superclusterOptions: this.superclusterOptions
-        };
+            extent: EXTENT,
+            scale: EXTENT / this.tileSize,
+            minZoom: this.minzoom,
+            maxZoom: this.maxzoom
+        }, this.options);
 
         var data = this._data;
         if (typeof data === 'string') {
             options.url = typeof window != 'undefined' ? urlResolve(window.location.href, data) : data;
         } else {
             options.data = JSON.stringify(data);
         }
-        this.workerID = this.dispatcher.send('parse geojson', options, function(err) {
+        this.workerID = this.dispatcher.send(this.type + '.parse', options, function(err) {
+            // TODO: not quite sure if we need this anymore
             this._loaded = true;
```

Comment:
I think we still need a `this._loaded` flag and `this.loaded` method (though not necessarily by those names) to implement #1715